### PR TITLE
Redesign for v 0.10.0

### PR DESCRIPTION
**Pull request by gusty**
This is an important re-write of the library:

Main points are:

  - Internally using mainly codecs (more efficient and less boilerplate in source code).
  - Ability to switch to different implementations, even at the same time. Transformation code is independent and thus re-usable, this allows to use several Json implementations, or even non Json encodings.
  - Unify Codec types, no more ConcreteCodec vs tupled codecs, just a simple 4 params Codec type.
  - A new and recommended Codec syntax for records and DUs trough an Applicative CE.
  - Internal caching of codecs.
  - Ability to workaround codecs for interfaces.

Minor points:
 - TimeSpan support
 - NonEmpties: List, Set and Map support
 - "Generic Map" support (Maps where keys are not strings).

**Tasks**

- [x] Ensure maximum compatibility with previous releases
- [x] Add deprecation warnings
- [x] Add codec cache
- [x] Add FSharpData
- [x] Add NewtonSoft
- [x] Add System.Json
- [x] Update System.Text.Json in order to use a single wrapper type
- [x] Re-implement lenses
- [x] Replace defaultof<_>s with error messages
- [x] Add tests for new features
- [x] Add tests for more than one overload per implementation
- [x] Review namespaces and modules
- [x] Update docs
- [x] Review dependencies, select the right version

### CODE REVIEW

File: appveyor.yml [hunk 1]
```diff
@@ -14,10 +14,7 @@ build_script:
   - cmd: echo vs %VersionSuffix%"
   - cmd: build.bat
 test_script:
-  - cmd: dotnet run --project .\test\Tests.SystemJson\Tests.SystemJson.fsproj -c Release -f net461
-  - cmd: dotnet run --project .\test\Tests.FSharpData\Tests.FSharpData.fsproj -c Release -f net5
```
Comment (wallymathieu) [inline appveyor.yml:18]: better not forget this

File: .github/workflows/dotnetcore.yml [hunk 2]
```diff
@@ -19,35 +19,6 @@ jobs:
         dotnet-version: 5.0.402
     - name: Install dependencies
       run: dotnet restore
-    - name: Build
```
Comment (wallymathieu) [inline .github/workflows/dotnetcore.yml:22]: better not forget this

File: appveyor.yml (hunk 1 above)
Comment (gusty) [reply appveyor.yml:18]: Don't worry, the plan is to revert these CI files when I have all 4 libs working.

File: .github/workflows/dotnetcore.yml (hunk 2 above)
Comment (gusty) [reply .github/workflows/dotnetcore.yml:22]: Don't worry, the plan is to revert these CI files when I have all 4 libs working.

File: .github/workflows/dotnetcore.yml [hunk 3]
```diff
@@ -19,35 +19,9 @@ jobs:
...
-    - name: Test SystemJson
-      run: dotnet run --no-build --project ./test/Tests.SystemJson/Tests.SystemJson.fsproj -c Release -f net5
-    - name: Test FSharpData
-      run: dotnet run --no-build --project ./test/Tests.FSharpData/Tests.FSharpData.fsproj -c Release -f net5
-    - name: Test NewtonsoftJson
-      run: dotnet run --no-build --project ./test/Tests.NewtonsoftJson/Tests.NewtonsoftJson.fsproj -c Release -f net5
-    - name: Test SystemTextJson
-      run: dotnet run --no-build --project ./test/Tests.SystemTextJson/Tests.SystemTextJson.fsproj -c Release -f net5
-
-  docs:
```
Comment (wallymathieu) [inline .github/workflows/dotnetcore.yml:33]: docs ...

File: test/Tests.SystemTextJson/Tests.SystemTextJson.fsproj [hunk 4]
```diff
@@ -10,7 +10,6 @@
         <Compile Include="../Tests/Tests.fs" />
     </ItemGroup>
     <ItemGroup>
-        <PackageReference Update="FSharp.Core" Version="4.6.2" />
```
Comment (wallymathieu) [inline test/Tests.SystemTextJson/Tests.SystemTextJson.fsproj:13]: should use 5.0.1

**Review by wallymathieu — APPROVED** [review]
This will make it significantly easier for others to contribute! Great work @gusty !

**Pull request merged**